cdc/sink/cdclog/s3.go

```diff
@@ -0,0 +1,599 @@
+			return ctx.Err()
+		case s.tableBuffers[hash].dataCh <- row:
+			s.tableBuffers[hash].lock.RLock()
+			s.tableBuffers[hash].sendSize += row.ApproximateSize
+			if s.tableBuffers[hash].sendSize > maxNotifySize {
+				// trigger flush when a table has maxNotifySize
+				shouldFlush = true
+			}
+			s.tableBuffers[hash].sendEvents += 1
+			s.tableBuffers[hash].lock.RUnlock()
+		}
+	}
+	if shouldFlush {
+		s.notifyChan <- struct{}{}
+	}
+	return nil
+}
+
+func (s *s3Sink) flushTableBuffers(ctx context.Context) error {
+	var wg sync.WaitGroup
+	errCh := make(chan error, len(s.tableBuffers))
+	for _, tb := range s.tableBuffers {
+		wg.Add(1)
+		tbReplica := tb
+		// TODO use a fixed worker pool
+		go func(ctx context.Context, tb *tableBuffer) {
+			defer wg.Done()
+			log.Info("[FlushRowChangedEvents] flush specify row changed event",
+				zap.Int64("table", tb.tableID),
+				zap.Int64("event size", tb.sendEvents))
+			errCh <- tb.flush(ctx, s)
+		}(ctx, tbReplica)
+	}
+	wg.Wait()
+	close(errCh)
+	for err := range errCh {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *s3Sink) FlushRowChangedEvents(ctx context.Context, resolvedTs uint64) error {
+	// we should flush all events before resolvedTs, there are two kind of flush policy
+	// 1. flush row events to a s3 chunk: if the event size is not enough,
+	//    TODO: when cdc crashed, we should repair these chunks to a complete file
+	// 2. flush row events to a complete s3 file: if the event size is enough
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.notifyChan:
+		return s.flushTableBuffers(ctx)
+
+	case <-time.After(defaultFlushRowChangedEventDuration):
+		// cannot accumulate enough row events in 10 second
+		// flush all tables' row events to s3
+		return s.flushTableBuffers(ctx)
+	}
+}
+
+// EmitCheckpointTs update the global resolved ts in log meta
+// sleep 5 seconds to avoid update too frequently
+func (s *s3Sink) EmitCheckpointTs(ctx context.Context, ts uint64) error {
+	s.logMeta.GlobalResolvedTS = ts
+	data, err := s.logMeta.Marshal()
+	if err != nil {
+		return errors.Annotate(err, "marshal meta to json failed")
+	}
+	return s.putObjectWithKey(ctx, s.prefix+logMetaFile, data)
+}
+
+// EmitDDLEvent write ddl event to S3 directory, all events split by '\n'
+// Because S3 doesn't support append-like write.
+// we choose a hack way to read origin file then write in place.
+func (s *s3Sink) EmitDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	switch ddl.Type {
+	case parsemodel.ActionCreateTable:
+		s.logMeta.Names[ddl.TableInfo.TableID] = makeSpecifyTableName(ddl.TableInfo.Schema, ddl.TableInfo.Table)
+	case parsemodel.ActionRenameTable:
+		delete(s.logMeta.Names, ddl.PreTableInfo.TableID)
+		s.logMeta.Names[ddl.TableInfo.TableID] = makeSpecifyTableName(ddl.TableInfo.Schema, ddl.TableInfo.Table)
+	}
+	data, err := ddl.ToProtoBuf().Marshal()
+	if err != nil {
+		return err
+	}
+	listing, err := s.ListObject(ddlEventsDir, 1)
```

Comment:
`MaxKeys means the response of `ListObjectsV2Input` returns at most one key, do we need to query the latest file?

Reply:
According to [Document](https://docs.aws.amazon.com/AmazonS3/latest/dev/ListingKeysUsingAPIs.html),` List results are always returned in UTF-8 binary order.`

So I choose `max.Long - commitTs` as suffix of ddl log file. this guarantee the ListObject will return the latest file.